src/osd/pg_scrubber.h

```diff
@@ -0,0 +1,631 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <cassert>
+#include <chrono>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "PG.h"
+#include "ScrubStore.h"
+#include "scrub_machine_lstnr.h"
+#include "scrubber_common.h"
+
+class Callback;
+
+namespace Scrub {
+class ScrubMachine;
+struct BuildMap;
+
+
+/**
+  ReplicaReservations is an RAII wrapper of the process of reserving/freeing scrub
+  resources at the replicas.
+
+  When constructed - sends reservation requests to the acting_set.
+  Handles arriving responses:
+  - a 'grant' is marked by reducing the 'pending_' counter. A list of reserved
+    resources is maintained.
+  - a rejection triggers a "couldn't acquire the replicas' scrub resources" event.
+    All "grants" arriving after the first rejection are denied, and cause an immediate
+    release of the resource.
+
+  Upon destruction - the acting-set (actually - those of the acting-set that have
+  'granted' our request - are freed.
+*/
+class ReplicaReservations {
+  using OrigSet =
+    decltype(std::declval<PG>().get_actingset());  // as I plan to change this type
```

Comment:
Unnecessary comment.  In general, we don't need to leave comments explaining things that should be clear to a C++ user.

Reply:
I will shorten this one. Still - I think our code suffers from too few comments. Even if the code is "obvious", I usually do not want to be forced to read every function when trying to understand what a piece of code does.